geo/geo*fn: error when comparing two geometries of different SRIDs

otan opened the pull request:
When operating on more than one geo type, we must ensure that if we compare
SRIDs that mismatch, we report an error. We've done this for all binary
predicates implemented so far in Geometry and Geography.

Furthermore, this PR adds an interface for Geospatial types. It makes
things a little bit neater, and we can extend this as we go along.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/48522)
<!-- Reviewable:end -->

otan committed:
geo/geo*fn: error when comparing two geometries of different SRIDs

When operating on more than one geo type, we must ensure that if we compare
SRIDs that mismatch, we report an error. We've done this for all binary
predicates implemented so far in Geometry and Geography.

Furthermore, this PR adds an interface for Geospatial types. It makes
things a little bit neater, and we can extend this as we go along.

Release note: None

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 16 of 16 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48522#-:-M6oKfmn-jSM1j8Yc4OS:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

otan commented:
bors r=sumeerbhola
